src/mlir/dialect/VeronaOps.cc

```diff
@@ -111,6 +133,32 @@ static LogicalResult verify(verona::ClassOp classOp)
   return success();
 }
 
+static LogicalResult verify(verona::WhileOp whileOp)
+{
+  auto& body = whileOp.body();
+
+  // While blocks must have a condition on the first block
+  // with a LoopExit operation.
+  bool hasExit = false;
+  for (Operation& op : body.getBlocks().front())
+  {
+    if (verona::LoopExitOp exit_op = dyn_cast<verona::LoopExitOp>(op))
+    {
+      hasExit = true;
+      break;
+    }
+  }
+
+  // And the terminator of the last block must be a continue.
+  bool hasCont =
+    isa<verona::ContinueOp>(body.getBlocks().back().getTerminator());
```

Comment:
This seems very restrictive, and I don't really see what it's trying to prevent.

The following should be valid IMO:
```
verona.while {
    jump ^bb1
^bb0:
    verona.continue
^bb1:
    jump ^bb0
}
```

Reply:
That's certainly possible, and we even have a test for that. 

The line is only trying to add a default continue on the last block if none was added by the code. So, if the code is:
```
while (...) {
  if (...) continue;
  break;
}
```

You don't add an additional `continue` after the break. But most loops will look like:
```
while (...) {
  if (...) ...;
}
```
And no terminator will be added to the last block, thus needing a continue.

Reply:
> That's certainly possible, and we even have a test for that.

Do we? I don't see any test where the last block does not end with a `continue`. Only a test which has a `continue` in both in the middle and at the end.

> The line is only trying to add a default continue on the last block if none was added by the code

That line doesn't add anything as far as I can tell. It only ensures it exists. It's the builder that adds the missing `continue` (and that's totally fine).

Here's another example that, I believe would be rejected by this verifier, even though it shouldn't:
```
while (...) {
    return;
}
```

Reply:
> ```
> while (...) {
>     return;
> }
> ```

Ah, I see what you mean! The back-branch example earlier threw me off the track. Just checking for all basic blocks having terminators should be enough.

The version with `cond_br` instead of `loop_exit` has a tail block that was the `continue` and why the check remained like this.